### PR TITLE
[`Refactor Attention mask handling`] Moves attention mask processing to the Attention class

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
This is more aligned with our philosophy, but also simplifies and will simplify things.
Will help a lot with the static cache.

The only way to share the mask is to call `LlamaAttention` but if you have a better way I'll update it! 
This makes the attention class self contained, which is also pretty convenient for testing. 
Ran the slow test without fa2 will run them again on dgx once approved. 

cc @patrickvonplaten for visibility

### CODE REVIEW

**Review by gante — COMMENTED** [review]
LGTM (yay, fewer config-dependent if/elses 🙌 )

BTW, for retrocompatibility, we may want to check whether the attention masks are 4D before expanding in the attention classes. As we've learned with the Cache refactor, other repos might rely on the interface of these internal classes, and this is technically an interface change (4D attention mask input -> 2D attention mask input).

**Comment by ArthurZucker** [comment]
We added support for 4d attention mask inside the converter so should be alright but yeah will check related issues!

File: src/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -358,11 +360,6 @@ def forward(
         use_cache: bool = False,
         **kwargs,
     ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
-        if "padding_mask" in kwargs:
-            warnings.warn(
-                "Passing `padding_mask` is deprecated and will be removed in v4.37. Please make sure use `attention_mask` instead.`"
-            )
-
```
Comment (LysandreJik) [inline src/transformers/models/llama/modeling_llama.py:365]: :wave:

File: src/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -285,6 +285,8 @@ def repeat_kv(hidden_states: torch.Tensor, n_rep: int) -> torch.Tensor:
 class LlamaAttention(nn.Module):
     """Multi-headed attention from 'Attention Is All You Need' paper"""
 
+    cached_mask = None
```
Comment (LysandreJik) [inline src/transformers/models/llama/modeling_llama.py:288]: as seen with you offline, this cannot work as-is due to sharing across model instances

Comment (ArthurZucker) [reply src/transformers/models/llama/modeling_llama.py:288]: Yes will either not cache it at the class level but instance level a tril. Or pass it as kwargs. Jax does not seem to care so should not be too bad

Comment (patrickvonplaten) [reply src/transformers/models/llama/modeling_llama.py:288]: Why can't we pass the `attention_mask` just into the `cache.update(...)` function?

**Comment by patrickvonplaten** [comment]
Can you specify how this helps with the static cache?

The static cache should also work with the attention_mask being passed at every forward call (it'll always have the same shape). I don't think it's a good idea to have the `attention_mask` be a class variable.

**Comment by ArthurZucker** [comment]
It will not be a class variable forgot to update but I'll follow what we do with jax.
This will help as the cache length is different from the number of tokens that are seen which you get when you are in the attention layer.

**Comment by ArthurZucker** [comment]
Can give more details but basically new cache + attention was not behaving properly. This is gonna be my priority this week anyway!

File: src/transformers/models/llama/modeling_llama.py (hunk 2 above)
Comment (ArthurZucker) [reply src/transformers/models/llama/modeling_llama.py:288]: I'll check that as well, but that doesn't help for all type of attention we have which need a pre-processed mask, will work after the pre-processing tho
